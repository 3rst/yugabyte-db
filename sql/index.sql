\unset ECHO
\i test_setup.sql

-- $Id$

SELECT plan(63);
--SELECT * FROM no_plan();

-- This will be rolled back. :-)
SET client_min_messages = warning;
CREATE TABLE public.sometab(
    id    INT NOT NULL PRIMARY KEY,
    name  TEXT DEFAULT '' UNIQUE,
    numb  NUMERIC(10, 2),
    myint NUMERIC(8)
);
CREATE INDEX idx_foo ON public.sometab(name);
CREATE INDEX idx_bar ON public.sometab(name, numb);
CREATE INDEX idx_baz ON public.sometab(LOWER(name));
RESET client_min_messages;

/****************************************************************************/
-- Test has_index().

SELECT * FROM check_test(
    has_index( 'public', 'sometab', 'idx_foo', 'name', 'whatever' ),
    true,
    'has_index() single column',
    'whatever',
    ''
);

SELECT * FROM check_test(
    has_index( 'public', 'sometab', 'idx_foo', 'name'::name ),
    true,
    'has_index() single column no desc',
    'Index "idx_foo" should exist',
    ''
);

SELECT * FROM check_test(
    has_index( 'public', 'sometab', 'idx_bar', ARRAY['name', 'numb'], 'whatever' ),
    true,
    'has_index() multi-column',
    'whatever',
    ''
);

SELECT * FROM check_test(
    has_index( 'public', 'sometab', 'idx_bar', ARRAY['name', 'numb'] ),
    true,
    'has_index() multi-column no desc',
    'Index "idx_bar" should exist',
    ''
);

SELECT * FROM check_test(
    has_index( 'public', 'sometab', 'idx_baz', 'LOWER(name)', 'whatever' ),
    true,
    'has_index() functional',
    'whatever',
    ''
);

SELECT * FROM check_test(
    has_index( 'public', 'sometab', 'idx_baz', 'whatever' ),
    true,
    'has_index() no cols',
    'whatever',
    ''
);

SELECT * FROM check_test(
    has_index( 'public', 'sometab', 'idx_baz'::name ),
    true,
    'has_index() no cols no desc',
    'Index "idx_baz" should exist',
    ''
);

SELECT * FROM check_test(
    has_index( 'sometab', 'idx_foo', 'name', 'whatever' ),
    true,
    'has_index() no schema single column',
    'whatever',
    ''
);

SELECT * FROM check_test(
    has_index( 'sometab', 'idx_foo', 'name'::name ),
    true,
    'has_index() no schema single column no desc',
    'Index "idx_foo" should exist',
    ''
);

SELECT * FROM check_test(
    has_index( 'sometab', 'idx_bar', ARRAY['name', 'numb'], 'whatever' ),
    true,
    'has_index() no schema multi-column',
    'whatever',
    ''
);

SELECT * FROM check_test(
    has_index( 'sometab', 'idx_bar', ARRAY['name', 'numb'] ),
    true,
    'has_index() no schema multi-column no desc',
    'Index "idx_bar" should exist',
    ''
);

SELECT * FROM check_test(
    has_index( 'sometab', 'idx_baz', 'LOWER(name)', 'whatever' ),
    true,
    'has_index() no schema functional',
    'whatever',
    ''
);

SELECT * FROM check_test(
    has_index( 'sometab', 'idx_baz', 'LOWER(name)' ),
    true,
    'has_index() no schema functional no desc',
    'Index "idx_baz" should exist',
    ''
);

SELECT * FROM check_test(
    has_index( 'sometab', 'idx_baz', 'whatever' ),
    true,
    'has_index() no schema or cols',
    'whatever',
    ''
);

SELECT * FROM check_test(
    has_index( 'sometab', 'idx_baz' ),
    true,
    'has_index() no schema or cols or desc',
    'Index "idx_baz" should exist',
    ''
);

-- Check failure diagnostics.
SELECT * FROM check_test(
    has_index( 'public', 'sometab', 'blah', ARRAY['name', 'numb'], 'whatever' ),
    false,
    'has_index() missing',
    'whatever',
    'Index "blah" ON public.sometab not found'
);

SELECT * FROM check_test(
    has_index( 'public', 'sometab', 'idx_bar', ARRAY['name', 'id'], 'whatever' ),
    false,
    'has_index() invalid',
    'whatever',
    '       have: "idx_bar" ON public.sometab(name, numb)
        want: "idx_bar" ON public.sometab(name, id)'
);

SELECT * FROM check_test(
    has_index( 'sometab', 'blah', ARRAY['name', 'numb'], 'whatever' ),
    false,
    'has_index() missing no schema',
    'whatever',
    'Index "blah" ON sometab not found'
);

SELECT * FROM check_test(
    has_index( 'sometab', 'idx_bar', ARRAY['name', 'id'], 'whatever' ),
    false,
    'has_index() invalid no schema',
    'whatever',
    '       have: "idx_bar" ON sometab(name, numb)
        want: "idx_bar" ON sometab(name, id)'
);

SELECT * FROM check_test(
    has_index( 'public', 'sometab', 'idx_baz', 'LOWER(wank)', 'whatever' ),
    false,
    'has_index() functional fail',
    'whatever',
    '       have: "idx_baz" ON public.sometab(lower(name))
        want: "idx_baz" ON public.sometab(lower(wank))'
);

SELECT * FROM check_test(
    has_index( 'sometab', 'idx_baz', 'LOWER(wank)', 'whatever' ),
    false,
    'has_index() functional fail no schema',
    'whatever',
    '       have: "idx_baz" ON sometab(lower(name))
        want: "idx_baz" ON sometab(lower(wank))'
);


/****************************************************************************/
-- Finish the tests and clean up.
SELECT * FROM finish();
ROLLBACK;
