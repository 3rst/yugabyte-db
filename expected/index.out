\unset ECHO
1..153
ok 1 - has_index() single column should pass
ok 2 - has_index() single column should have the proper description
ok 3 - has_index() single column should have the proper diagnostics
ok 4 - has_index() single column no desc should pass
ok 5 - has_index() single column no desc should have the proper description
ok 6 - has_index() single column no desc should have the proper diagnostics
ok 7 - has_index() multi-column should pass
ok 8 - has_index() multi-column should have the proper description
ok 9 - has_index() multi-column should have the proper diagnostics
ok 10 - has_index() multi-column no desc should pass
ok 11 - has_index() multi-column no desc should have the proper description
ok 12 - has_index() multi-column no desc should have the proper diagnostics
ok 13 - has_index() functional should pass
ok 14 - has_index() functional should have the proper description
ok 15 - has_index() functional should have the proper diagnostics
ok 16 - has_index() no cols should pass
ok 17 - has_index() no cols should have the proper description
ok 18 - has_index() no cols should have the proper diagnostics
ok 19 - has_index() no cols no desc should pass
ok 20 - has_index() no cols no desc should have the proper description
ok 21 - has_index() no cols no desc should have the proper diagnostics
ok 22 - has_index() no schema single column should pass
ok 23 - has_index() no schema single column should have the proper description
ok 24 - has_index() no schema single column should have the proper diagnostics
ok 25 - has_index() no schema single column no desc should pass
ok 26 - has_index() no schema single column no desc should have the proper description
ok 27 - has_index() no schema single column no desc should have the proper diagnostics
ok 28 - has_index() no schema multi-column should pass
ok 29 - has_index() no schema multi-column should have the proper description
ok 30 - has_index() no schema multi-column should have the proper diagnostics
ok 31 - has_index() no schema multi-column no desc should pass
ok 32 - has_index() no schema multi-column no desc should have the proper description
ok 33 - has_index() no schema multi-column no desc should have the proper diagnostics
ok 34 - has_index() no schema functional should pass
ok 35 - has_index() no schema functional should have the proper description
ok 36 - has_index() no schema functional should have the proper diagnostics
ok 37 - has_index() no schema functional no desc should pass
ok 38 - has_index() no schema functional no desc should have the proper description
ok 39 - has_index() no schema functional no desc should have the proper diagnostics
ok 40 - has_index() no schema or cols should pass
ok 41 - has_index() no schema or cols should have the proper description
ok 42 - has_index() no schema or cols should have the proper diagnostics
ok 43 - has_index() no schema or cols or desc should pass
ok 44 - has_index() no schema or cols or desc should have the proper description
ok 45 - has_index() no schema or cols or desc should have the proper diagnostics
ok 46 - has_index() missing should fail
ok 47 - has_index() missing should have the proper description
ok 48 - has_index() missing should have the proper diagnostics
ok 49 - has_index() invalid should fail
ok 50 - has_index() invalid should have the proper description
ok 51 - has_index() invalid should have the proper diagnostics
ok 52 - has_index() missing no schema should fail
ok 53 - has_index() missing no schema should have the proper description
ok 54 - has_index() missing no schema should have the proper diagnostics
ok 55 - has_index() invalid no schema should fail
ok 56 - has_index() invalid no schema should have the proper description
ok 57 - has_index() invalid no schema should have the proper diagnostics
ok 58 - has_index() functional fail should fail
ok 59 - has_index() functional fail should have the proper description
ok 60 - has_index() functional fail should have the proper diagnostics
ok 61 - has_index() functional fail no schema should fail
ok 62 - has_index() functional fail no schema should have the proper description
ok 63 - has_index() functional fail no schema should have the proper diagnostics
ok 64 - index_is_unique() should pass
ok 65 - index_is_unique() should have the proper description
ok 66 - index_is_unique() should have the proper diagnostics
ok 67 - index_is_unique() no desc should pass
ok 68 - index_is_unique() no desc should have the proper description
ok 69 - index_is_unique() no desc should have the proper diagnostics
ok 70 - index_is_unique() no schema should pass
ok 71 - index_is_unique() no schema should have the proper description
ok 72 - index_is_unique() no schema should have the proper diagnostics
ok 73 - index_is_unique() index only should pass
ok 74 - index_is_unique() index only should have the proper description
ok 75 - index_is_unique() index only should have the proper diagnostics
ok 76 - index_is_unique() on pk should pass
ok 77 - index_is_unique() on pk should have the proper description
ok 78 - index_is_unique() on pk should have the proper diagnostics
ok 79 - index_is_unique() on pk no desc should pass
ok 80 - index_is_unique() on pk no desc should have the proper description
ok 81 - index_is_unique() on pk no desc should have the proper diagnostics
ok 82 - index_is_unique() on pk no schema should pass
ok 83 - index_is_unique() on pk no schema should have the proper description
ok 84 - index_is_unique() on pk no schema should have the proper diagnostics
ok 85 - index_is_unique() on pk index only should pass
ok 86 - index_is_unique() on pk index only should have the proper description
ok 87 - index_is_unique() on pk index only should have the proper diagnostics
ok 88 - index_is_unique() fail should fail
ok 89 - index_is_unique() fail should have the proper description
ok 90 - index_is_unique() fail should have the proper diagnostics
ok 91 - index_is_unique() fail no desc should fail
ok 92 - index_is_unique() fail no desc should have the proper description
ok 93 - index_is_unique() fail no desc should have the proper diagnostics
ok 94 - index_is_unique() fail no schema should fail
ok 95 - index_is_unique() fail no schema should have the proper description
ok 96 - index_is_unique() fail no schema should have the proper diagnostics
ok 97 - index_is_unique() fail index only should fail
ok 98 - index_is_unique() fail index only should have the proper description
ok 99 - index_is_unique() fail index only should have the proper diagnostics
ok 100 - index_is_unique() no such index should fail
ok 101 - index_is_unique() no such index should have the proper description
ok 102 - index_is_unique() no such index should have the proper diagnostics
ok 103 - index_is_primary() should pass
ok 104 - index_is_primary() should have the proper description
ok 105 - index_is_primary() should have the proper diagnostics
ok 106 - index_is_primary() no desc should pass
ok 107 - index_is_primary() no desc should have the proper description
ok 108 - index_is_primary() no desc should have the proper diagnostics
ok 109 - index_is_primary() no schema should pass
ok 110 - index_is_primary() no schema should have the proper description
ok 111 - index_is_primary() no schema should have the proper diagnostics
ok 112 - index_is_primary() index only should pass
ok 113 - index_is_primary() index only should have the proper description
ok 114 - index_is_primary() index only should have the proper diagnostics
ok 115 - index_is_primary() fail should fail
ok 116 - index_is_primary() fail should have the proper description
ok 117 - index_is_primary() fail should have the proper diagnostics
ok 118 - index_is_primary() fail no desc should fail
ok 119 - index_is_primary() fail no desc should have the proper description
ok 120 - index_is_primary() fail no desc should have the proper diagnostics
ok 121 - index_is_primary() fail no schema should fail
ok 122 - index_is_primary() fail no schema should have the proper description
ok 123 - index_is_primary() fail no schema should have the proper diagnostics
ok 124 - index_is_primary() fail index only should fail
ok 125 - index_is_primary() fail index only should have the proper description
ok 126 - index_is_primary() fail index only should have the proper diagnostics
ok 127 - index_is_primary() no such index should fail
ok 128 - index_is_primary() no such index should have the proper description
ok 129 - index_is_primary() no such index should have the proper diagnostics
ok 130 - is_clustered() fail should fail
ok 131 - is_clustered() fail should have the proper description
ok 132 - is_clustered() fail should have the proper diagnostics
ok 133 - is_clustered() fail no desc should fail
ok 134 - is_clustered() fail no desc should have the proper description
ok 135 - is_clustered() fail no desc should have the proper diagnostics
ok 136 - is_clustered() fail no schema should fail
ok 137 - is_clustered() fail no schema should have the proper description
ok 138 - is_clustered() fail no schema should have the proper diagnostics
ok 139 - is_clustered() fail index only should fail
ok 140 - is_clustered() fail index only should have the proper description
ok 141 - is_clustered() fail index only should have the proper diagnostics
ok 142 - is_clustered() should pass
ok 143 - is_clustered() should have the proper description
ok 144 - is_clustered() should have the proper diagnostics
ok 145 - is_clustered() no desc should pass
ok 146 - is_clustered() no desc should have the proper description
ok 147 - is_clustered() no desc should have the proper diagnostics
ok 148 - is_clustered() no schema should pass
ok 149 - is_clustered() no schema should have the proper description
ok 150 - is_clustered() no schema should have the proper diagnostics
ok 151 - is_clustered() index only should pass
ok 152 - is_clustered() index only should have the proper description
ok 153 - is_clustered() index only should have the proper diagnostics
